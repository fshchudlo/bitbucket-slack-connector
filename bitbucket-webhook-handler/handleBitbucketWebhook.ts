import * as useCases from "./use-cases";
import { SlackAPIAdapter } from "./ports/SlackAPIAdapter";
import { BitbucketAPIAdapter } from "./ports/BitbucketAPIAdapter";
import { BitbucketNotification } from "../bitbucket-payload-types";
import { WebhookHandlerConfig } from "./webhookHandlerConfig";
import { provisionPullRequestChannel } from "./provisionPullRequestChannel";

export default async function handleBitbucketWebhook(payload: BitbucketNotification, slackAPI: SlackAPIAdapter, bitbucketAPI: BitbucketAPIAdapter, config: WebhookHandlerConfig) {
    const eventKey = payload.eventKey;
    const channelInfo = await provisionPullRequestChannel(slackAPI, bitbucketAPI, payload, config);

    switch (eventKey) {
        case "pr:opened":
            await useCases.inviteParticipantsAndSetChannelBookmark(payload, slackAPI, config.defaultChannelParticipants, channelInfo.id);
            await useCases.tryBroadcastMessageAboutOpenedPR(payload, slackAPI, config.getOpenedPRBroadcastChannelId(payload));
            break;
        case "pr:modified":
            await useCases.sendMessageAboutPRModification(payload, slackAPI, channelInfo.id);
            break;
        case "pr:reviewer:updated":
            await useCases.updateChannelMembers(payload, slackAPI, channelInfo.id);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await useCases.sendMessageAboutReviewerAction(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:added":
            await useCases.sendMessageAboutAddedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:edited":
            await useCases.sendMessageAboutEditedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:deleted":
            await useCases.sendMessageAboutDeletedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:from_ref_updated":
            await useCases.sendMessageAboutNewCommit(payload, slackAPI, bitbucketAPI, channelInfo.id);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await useCases.sendCompletionMessageAndCloseTheChannel(payload, slackAPI, channelInfo.id);
            await useCases.tryBroadcastMessageAboutClosedPR(payload, slackAPI, config.getOpenedPRBroadcastChannelId(payload));
            break;
        default:
            throw new Error(`"${eventKey}" event key is unknown.`);
    }
}

