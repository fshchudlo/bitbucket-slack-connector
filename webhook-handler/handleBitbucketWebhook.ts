import {
    sendCompletionMessageAndCloseTheChannel,
    setChannelTopicAndInviteParticipants,
    sendMessageAboutAddedComment,
    updateChannelMembers,
    sendMessageAboutNewCommit,
    sendMessageAboutReviewerAction,
    sendMessageAboutDeletedComment, sendMessageAboutEditedComment
} from "./use-cases";
import { SlackAPIAdapter } from "./SlackAPIAdapter";
import { BitbucketGateway } from "./BitbucketGateway";
import { sendMessageAboutPRModification } from "./use-cases/sendMessageAboutPRModification";
import { BitbucketNotification } from "../typings";
import { buildChannelName } from "./slack-helpers";

export default async function handleBitbucketWebhook(payload: BitbucketNotification, slackAPI: SlackAPIAdapter, bitbucketGateway: BitbucketGateway, usePrivateChannels: boolean = true) {
    const eventKey = payload.eventKey;
    const channelInfo = await provisionPullRequestChannel(slackAPI, payload, usePrivateChannels);
    switch (eventKey) {
        case "pr:opened":
            await setChannelTopicAndInviteParticipants(payload, slackAPI, channelInfo);
            break;
        case "pr:modified":
            await sendMessageAboutPRModification(payload, slackAPI, channelInfo);
            break;
        case "pr:reviewer:updated":
            await updateChannelMembers(payload, slackAPI, channelInfo);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await sendMessageAboutReviewerAction(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:added":
            await sendMessageAboutAddedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:edited":
            await sendMessageAboutEditedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:deleted":
            await sendMessageAboutDeletedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:from_ref_updated":
            await sendMessageAboutNewCommit(payload, slackAPI, bitbucketGateway, channelInfo);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await sendCompletionMessageAndCloseTheChannel(payload, slackAPI, channelInfo);
            break;
        default:
            throw new Error(`"${eventKey}" event key is unknown.`);
    }
}

async function provisionPullRequestChannel(slackAPI: SlackAPIAdapter, payload: BitbucketNotification, usePrivateChannels: boolean) {
    return payload.eventKey == "pr:opened" ? await slackAPI.createChannel({
        name: buildChannelName(payload.pullRequest),
        isPrivate: usePrivateChannels
    }) : await slackAPI.findChannel(buildChannelName(payload.pullRequest), true);
}
