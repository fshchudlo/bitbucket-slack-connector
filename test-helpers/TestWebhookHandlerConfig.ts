import { WebhookHandlerConfig } from "../bitbucket-webhook-handler/webhookHandlerConfig";

export const TestWebhookHandlerConfig: WebhookHandlerConfig = {
    getOpenedPRBroadcastChannelId(): string | null {
        return null;
    },
    usePrivateChannels: true
};