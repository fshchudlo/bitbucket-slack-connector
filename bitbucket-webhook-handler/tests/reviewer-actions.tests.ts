import SlackAdapterSnapshottingMock from "./mocks/SlackAdapterSnapshottingMock";
import TestPayloadBuilder from "./mocks/TestPayloadBuilder";
import handleBitbucketWebhook from "../handleBitbucketWebhook";
import { TestWebhookHandlerConfig } from "./mocks/TestWebhookHandlerConfig";

describe("handleBitbucketWebhook", () => {
    it("Should send message on PR approval/unapproval/needs work", async () => {
        const testSlackGateway = await new SlackAdapterSnapshottingMock().setupBasicChannel();


        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestNeedsWork(), testSlackGateway, testSlackGateway.testChannel, TestWebhookHandlerConfig);
        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestUnapproved(), testSlackGateway, testSlackGateway.testChannel, TestWebhookHandlerConfig);
        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestApproved(), testSlackGateway, testSlackGateway.testChannel, TestWebhookHandlerConfig);


        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });
});
