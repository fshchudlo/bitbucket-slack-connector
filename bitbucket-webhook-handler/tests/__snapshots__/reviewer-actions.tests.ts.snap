// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleBitbucketWebhook Should send message on PR approval/unapproval/needs work 1`] = `
{
  "addedBookmarks": [
    {
      "channelId": "12345",
      "emoji": ":git:",
      "link": "https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview",
      "title": "Review Pull Request",
    },
  ],
  "addedReactions": [],
  "archivedChannels": [],
  "createdChannels": [
    {
      "id": "12345",
      "isArchived": false,
      "name": "pr-test-proj-test-repository-1",
    },
  ],
  "invitesToChannels": [
    {
      "channelId": "12345",
      "force": true,
      "users": [
        "test.author@test.com",
        "test.reviewer1@test.com",
        "test.reviewer2@test.com",
      ],
    },
  ],
  "kicksFromChannels": [],
  "lookedUpUsers": [
    [
      "test.author@test.com",
      "test.reviewer1@test.com",
      "test.reviewer2@test.com",
    ],
  ],
  "searchedChannels": [
    {
      "channelName": "pr-test-proj-test-repository-1",
      "findPrivateChannels": true,
    },
    {
      "channelName": "pr-test-proj-test-repository-1",
      "findPrivateChannels": true,
    },
    {
      "channelName": "pr-test-proj-test-repository-1",
      "findPrivateChannels": true,
    },
  ],
  "searchedCommentSnapshots": [],
  "searchedPrOpenedBroadcastMessages": [],
  "sentMessages": [
    {
      "blocks": [
        {
          "text": {
            "text": "Test Author opened <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "type": "divider",
        },
        {
          "elements": [
            {
              "text": "Test pull request description",
              "type": "mrkdwn",
            },
          ],
          "type": "context",
        },
        {
          "type": "divider",
        },
        {
          "text": {
            "text": "Please <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|review the PR>",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "text": "Test Author opened <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>",
    },
    {
      "blocks": [
        {
          "text": {
            "text": ":traffic_light: Test Reviewer 1 requested changes for the <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "type": "divider",
        },
        {
          "elements": [
            {
              "text": ":large_yellow_circle:  Approved: 0 |  Requested changes: Test Reviewer 1 |  Not reviewed: Test Reviewer 2",
              "type": "mrkdwn",
            },
          ],
          "type": "context",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "text": ":traffic_light: Test Reviewer 1 requested changes for the <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
    },
    {
      "blocks": [
        {
          "text": {
            "text": ":traffic_light: Test Reviewer 1 unapproved <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "type": "divider",
        },
        {
          "elements": [
            {
              "text": ":large_yellow_circle:  Approved: 0 |  Not reviewed: Test Reviewer 1,Test Reviewer 2",
              "type": "mrkdwn",
            },
          ],
          "type": "context",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "text": ":traffic_light: Test Reviewer 1 unapproved <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
    },
    {
      "blocks": [
        {
          "text": {
            "text": ":traffic_light: Test Reviewer 1 approved <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "type": "divider",
        },
        {
          "elements": [
            {
              "text": ":large_yellow_circle:  Approved: Test Reviewer 1 |  Not reviewed: Test Reviewer 2",
              "type": "mrkdwn",
            },
          ],
          "type": "context",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "text": ":traffic_light: Test Reviewer 1 approved <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
    },
  ],
}
`;
